### PR TITLE
feat: portals trade execution

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR does what it says on the box and implements trade execution of Portals quote.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/7306

> What protocols, transaction types or contract interactions might be affected by this PR?

Low, but this is swapper so... high? Note this is an entirely new swapper and doesn't touch existing swappers logic, and is still under flag, but as with all features, should be considered production-ready as-is.

## Testing

Portals swaps are happy e2e:

- Native EVM asset <-> Native EVM asset yields no Portals quote since Portals isn't cross-chain
- Token <-> Native EVM asset swaps are happy in both directions
- You should be able to continue to the confirm step and execute the trade succesfully even without an approval granted for that token (i.e quote doesn't fail on approval granted, since you'll approve at confirm step, business as usual!)

### Engineering

- ☝🏽 

### Operations

- ☝🏽 

### Native Asset -> Token (notice the succesful gras estimations)

https://jam.dev/c/28f4686e-e6f7-41ba-bbe9-973bc433995c

### Token -> Native Asset (approval not yet granted, notice the 0$ gas as Portals is unable to estimate upstream)

https://jam.dev/c/14e4ca8f-8042-4544-99e5-b87f8f198089

### Token -> Token (approval granted, notice the succesful gas estimations)

https://jam.dev/c/6bd6836f-2eb0-4599-a723-b81345a5ab5f

### Native Asset <-> Native Asset

Unapplicable - Portals doesn't do cross-chain swaps

### CODE REVIEW

File: packages/swapper/src/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts [hunk 1]
```diff
@@ -74,7 +74,12 @@ export async function getTradeQuote(
     affiliateBps,
   )
 
-  const rate = getRate(quote).toString()
+  const rate = getRate({
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts:77]: See below - `getRate` moved to root swapper utils and generalized vs. taking a OneInch specific response - actual logic doesn't change.

File: packages/swapper/src/swappers/OneInchSwapper/utils/helpers.ts [hunk 2]
```diff
@@ -1,27 +1,16 @@
...
 
-export const getRate = (response: OneInchBaseResponse): BigNumber => {
-  const fromTokenAmountCryptoHuman = fromBaseUnit(
-    response.fromTokenAmount,
-    response.fromToken.decimals,
-  )
-  const toTokenAmountCryptoHuman = fromBaseUnit(response.toTokenAmount, response.toToken.decimals)
-  return bn(toTokenAmountCryptoHuman).div(fromTokenAmountCryptoHuman)
-}
-
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/OneInchSwapper/utils/helpers.ts:24]: Moved around, not removed

File: packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts [hunk 3]
```diff
@@ -144,9 +144,12 @@ export async function getPortalsTradeQuote(
       },
     } = portalsTradeOrderResponse
 
-    const rate = bn(buyAmountAfterFeesCryptoBaseUnit)
-      .div(input.sellAmountIncludingProtocolFeesCryptoBaseUnit)
-      .toString()
+    const rate = getRate({
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:147]: This was previously off - sell asset and buy asset may have different base units, we can't do that

Before:

Screenshot 2024-07-29 at 12 41 05
Screenshot 2024-07-29 at 12 41 37

After:

Screenshot 2024-07-29 at 12 41 05
Screenshot 2024-07-29 at 12 21 14

Comment (NeOMakinG) [reply packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:151]: Nice

**Review by NeOMakinG — APPROVED** [review]
Looks good so far, nothing to say on the code side, even without gas fees estimation the trade did pass, looks good!

File: packages/swapper/src/utils.ts [hunk 4]
```diff
@@ -191,3 +194,19 @@ export const checkEvmSwapStatus = async ({
...
+  buyAsset,
+}: {
+  sellAmountCryptoBaseUnit: string
+  buyAmountCryptoBaseUnit: string
+  sellAsset: Asset
+  buyAsset: Asset
+}): BigNumber => {
+  const sellAmountCryptoHuman = fromBaseUnit(sellAmountCryptoBaseUnit, sellAsset.precision)
+  const buyAmountCryptoHuman = fromBaseUnit(buyAmountCryptoBaseUnit, buyAsset.precision)
+  return bn(buyAmountCryptoHuman).div(sellAmountCryptoHuman)
```
Comment (kaladinlight) [inline packages/swapper/src/utils.ts:211]: nitpick - return `toFixed()` and remove `toFixed()` and `toString()` from portals and one inch getTradeQuote respectively

**Comment by kaladinlight** [comment]
Noticing gas fee is wildly off so we can look into that a bit more as a follow up:
https://arbiscan.io/tx/0x317c8a9a590e0db3bcae0cac5f4f47038aa10fc2f0a6d797fdeb5e2e4f73b283
image

**Review by kaladinlight — APPROVED** [review]
Couple minor notes, but signing off on this one

File: packages/swapper/src/utils.ts (hunk 4 above)
Comment (gomesalexandre) [reply packages/swapper/src/utils.ts]: fd57bc186c0fc99f05f446572882093b0a5b68a6

**Pull request merged**